database2/ffboltdb/blockio.go

```diff
@@ -0,0 +1,746 @@
+	//
+	// The serialized block location format is:
+	//
+	//  [0:4]  Block file (4 bytes)
+	//  [4:8]  File offset (4 bytes)
+	//  [8:12] Block length (4 bytes)
+	blockLocSize = 12
+)
+
+var (
+	// castagnoli houses the Catagnoli polynomial used for CRC-32 checksums.
+	castagnoli = crc32.MakeTable(crc32.Castagnoli)
+)
+
+// filer is an interface which acts very similar to a *os.File and is typically
+// implemented by it.  It exists so the test code can provide mock files for
+// properly testing corruption and file system issues.
+type filer interface {
+	io.Closer
+	io.WriterAt
+	io.ReaderAt
+	Truncate(size int64) error
+	Sync() error
+}
+
+// lockableFile represents a block file on disk that has been opened for either
+// read or read/write access.  It also contains a read-write mutex to support
+// multiple concurrent readers.
+type lockableFile struct {
+	sync.RWMutex
+	file filer
+}
+
+// writeCursor represents the current file and offset of the block file on disk
+// for performing all writes. It also contains a read-write mutex to support
+// multiple concurrent readers which can reuse the file handle.
+type writeCursor struct {
+	sync.RWMutex
+
+	// curFile is the current block file that will be appended to when
+	// writing new blocks.
+	curFile *lockableFile
+
+	// curFileNum is the current block file number and is used to allow
+	// readers to use the same open file handle.
+	curFileNum uint32
+
+	// curOffset is the offset in the current write block file where the
+	// next new block will be written.
+	curOffset uint32
+}
+
+// blockStore houses information used to handle reading and writing blocks (and
+// part of blocks) into flat files with support for multiple concurrent readers.
+type blockStore struct {
+	// network is the specific network to use in the flat files for each
+	// block.
+	network wire.BitcoinNet
+
+	// basePath is the base path used for the flat block files and metadata.
+	basePath string
+
+	// maxBlockFileSize is the maximum size for each file used to store
+	// blocks.  It is defined on the store so the whitebox tests can
+	// override the value.
+	maxBlockFileSize uint32
+
+	// The following fields are related to the flat files which hold the
+	// actual blocks.   The number of open files is limited by maxOpenFiles.
+	//
+	// obfMutex protects concurrent access to the openBlockFiles map.  It is
+	// a RWMutex so multiple readers can simultaneously access open files.
+	//
+	// openBlockFiles houses the open file handles for existing block files
+	// which have been opened read-only along with an individual RWMutex.
+	// This scheme allows multiple concurrent readers to the same file while
+	// preventing the file from closed out from under them.
+	//
+	// lruMutex protects concurrent access to the least recently used list
+	// and lookup map.
+	//
+	// openBlocksLRU tracks how the open files are refenced by pushing the
+	// least recently used files to end of the list.  When a file needs to
```

Comment:
However I'll reword it in terms of the most recently used since I can see how that might be worded in a confusing fashion.